### PR TITLE
Prometheus Remote Write Backend Support

### PR DESCRIPTION
**Pull request by dsmith3197**
# Description
This PR adds support for a Prometheus Remote Write backend that can be used as an alternative to the statsd backend. 

## Approach

The new `PromClient` is implemented with the same interface as the existing `StatsdClient`.  The two can now be used interchangeably thanks to a new `BackendClient` trait.  This allows us to swap out the sink logic while leaving everything else in tact.

`PromClient` batches the statsd PDUs it receives according the batch policy specified in the configuration (see below section).  It then transforms a batch of PDUs into a Prometheus `WriteRequest` protobuf, snappy compresses it, and then finally sends it to the specified endpoint.

## Configuration
Backends can now be one of either two types: `statsd` and `prom_remote_write`

### Statsd
A statsd backend is configured in the same way as before with one exception: a new `type` field is now required.
```
{
    "statsd": {
        "backends": {
            "my_backend": {
                "type": "statsd",
                "prefix": "new_prefix.",
                "suffix": ".new_suffix",
                "shard_map": [
                    "127.0.0.1:4040"
                ]
            }
        }
    }
}
```

### Prometheus Remote Write
A Prometheus Remote Write backend is configured similarly.  Notably, sharding, prefixing/suffixing, and filtering for these backends can be configured in the same manner ([code ref](https://github.com/lyft/statsrelay/pull/76/files#diff-fd818c5b76f368e88976d9400c382ba46b523ea761d56613948b8750e5f414b9R122)). 
```
{
    "statsd": {
        "backends": {
            "my_backend": {
                "type": "prom_remote_write",
                "prefix": "new_prefix.",
                "suffix": ".new_suffix",
                "shard_map": [
                    "http://127.0.0.1:4040/remote-write"
                ],
                "max_queue": 100000,
                "request_timeout_ms": 10000,
                "retry_delay_ms": 250,
                "batch_creation_timeout_ms": 500,
                "max_in_flight": 16,
            }
        }
    }
}
```

## Statsd PDU -> Prometheus `WriteRequest` Protobuf Transformation

For the most part, the transformation from a collection of statsd PDUs to a Prometheus `WriteRequest` protobuf is fairly straightforward, with one `timeseries` per metric.  Notably, the metric types are stored separately in the `metadata` field.

With the current implementation a PDU is dropped if it is unparsable.

### Prometheus `WriteRequest` Protobuf Extension
While a statsd line can include a sample rate, that is not true for a `WriteRequest` protobuf.  We added a new `sample_rate` field to the `Sample` protobuf message to propagate this data.

### CODE REVIEW

File: statsrelay/src/backend.rs [hunk 1]
```diff
@@ -1,33 +1,39 @@
...
+    conf: config::BackendConfig,
+    ring: Ring<ArcBackendClient>,
     input_filter: Option<RegexSet>,
     warning_log: AtomicU64,
     backend_sends: stats::Counter,
     backend_fails: stats::Counter,
 }
 
-impl StatsdBackend {
+impl Backend {
```
Comment (bearcage) [inline statsrelay/src/backend.rs:32]: I'm not _certain_ there's a cleaner factoring available here, but it does seem like this type has a couple responsibilities that might be reasonably layered/composed rather than belonging to one type -- constructing the correct backend from a given config, handling discovery updates, managing a consistent hashing ring of shards, tracking filters, and carrying around stats.

I know the stats just kind of have to be passed down, so that's not really separable, but I wonder if maybe the config -> backend bit could be pulled out to a free function this calls, and/or perhaps composed in a ShardedBackendClient type to manage the ring, leaving the layering something like `Backend(filters, stats, config, ShardedBackendClient(config, stats, ring, BackendClient(config)))`?

It's not immediately clear to me whether or not this is worth doing at the current stage, but I do think if we're expecting to further extend this type it runs the risk of becoming an all-in-one class with a lot of branching logic that's more difficult to test and compose. Layering it would hopefully make that outcome less likely.

File: statsrelay/src/backend_client/mod.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+pub trait BackendClient {
+    fn sender(&self) -> mpsc::Sender<Pdu>;
+
+    fn endpoint(&self) -> &str;
+
+    // Request the join handle for this client, which completes when all
+    // operations attempting to use this client are complete and client has shut
+    // down. Note that this is a destructive action, and a second call to a join
+    // handle will return None.
+    fn join_handle(&self) -> Option<JoinHandle<()>>;
```
Comment (bearcage) [inline statsrelay/src/backend_client/mod.rs:17]: Def don't feel obligated to do this one, I'm just picking at JoinHandle appearing in a return position on a trait method since IME that particular idiom can lead to pretty verbose shutdown code for callers, if they want to gracefully drain things with a timeout.

Is there more to do with this joinhandle than just waiting for it to complete? If not, you might consider making the interface something like `async fn shutdown(self);` (or return an enum indicating state) and doing the flush/wait internal to this function. I know it'd mean pulling in async_trait and paying the extra indirection cost in the implementation, but it might be a little bit nicer for callers to interact with since it would encode the destructiveness in the type (it takes self by ownership).

Minor related nit: did you want this to be a doc-comment so it shows up in the rustdoc for this trait?

File: statsrelay/src/backend_client/mod.rs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+use tokio::{sync::mpsc, task::JoinHandle};
+
+use crate::statsd_proto::Pdu;
+
+pub mod prom_client;
+pub mod statsd_client;
+
+pub trait BackendClient {
+    fn sender(&self) -> mpsc::Sender<Pdu>;
```
Comment (bearcage) [inline statsrelay/src/backend_client/mod.rs:9]: Would it be worth factoring this as `send(&self, pdu: Pdu)` rather than an accessor for the sender, to avoid leaking the kind of sender it is through the abstraction?

File: statsrelay/src/backend_client/prom_client.rs [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+    inner: Arc<Mutex<PromClientInner>>,
+    sender: mpsc::Sender<Pdu>,
+}
+
+impl PromClient {
+    pub fn new(scope: stats::Scope, endpoint: &str, config: PromBackendClientConfig) -> Self {
+        let inner = Client::builder()
+            .timeout(Duration::from_millis(config.request_timeout_ms))
+            .build()
+            .expect("failed to build client");
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:64]: Is there a case in which a caller could recover from a failure to construct this (e.g. by falling back to another backend or something)? In other words, should this return a Result<Self> rather than panicking on construction?

If not, please add a doc-comment to this method to note that it panics on a failure to construct the backing client from a config.

File: statsrelay/src/backend_client/prom_client.rs [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+}
+
+async fn client_task<T: PromWriteRequestClient + Send + Sync + 'static>(
+    stats: Stats,
+    client: Arc<T>,
+    mut rx: mpsc::Receiver<Pdu>,
+    trip: Tripwire,
+    config: PromBackendClientConfig,
+) {
+    let semaphore = Arc::new(Semaphore::new(config.max_in_flight));
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:169]: Using max_in_flight at this layer means that config field is defining the maximum number of requests in flight _per client task_, correct? How does this interact with the shard ring of clients in the Backend type?

File: statsrelay/src/backend_client/prom_client.rs [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+                if let Err(e) = res {
+                    stats.requests_fail.inc();
+                    stats.samples_send_fail.inc_by(samples_in_batch);
+                    warn!(
+                        "prometheus remote write request failed, dropping samples: {:?}",
+                        e
+                    );
+                }
+
+                drop(permit);
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:243]: Minor nit: I'd recommend sticking a comment above these drops to indicate that they're here to force the move-into-closure. IME the rules around what gets closed over is a stumbling block for folks new to rust, so you can do them a favor when they read your code by tacking on a one-liner comment noting why the drops are explicit.

We'll need to do less of this as more of our team gets comfortable with rust, but for the time being IMO it's nice to try to make the gotchas more approachable.

File: statsrelay/src/backend_client/prom_client.rs [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+                    .bytes_sent_total
+                    .inc_by(compressed_write_request.len() as f64);
+
+                // Attempt each request at least once
+                let mut res = client
+                    .send_write_request(compressed_write_request.clone())
+                    .await;
+                // And retry only if the tripwire has not yet been triggered.
+                while should_retry(&res) {
+                    sleep(Duration::from_millis(retry_delay_ms)).await;
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:222]: Don't have to do it right now, but some jitter and/or backoff might be a nice future addition to the retry mechanism. You might also be able to incorporate a more general retry mechanism at a higher layer in this stack -- say by wrapping at the `BackendClient` layer with a `RetryingBackendClient : BackendClient` or similar -- to avoid having to implement it separately in each backend that's able to fail.

File: statsrelay/src/backend_client/prom_client.rs [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+    rx: &mut mpsc::Receiver<Pdu>,
+    timeout: Duration,
+    max_batch_size: usize,
+) -> (Vec<Pdu>, bool) {
+    let mut batch = Vec::<Pdu>::with_capacity(max_batch_size);
+    let sleep = sleep(timeout);
+    tokio::pin!(sleep);
+    loop {
+        select! {
+            opdu = rx.recv() => {
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:267]: nit: what's `opdu` stand for? Original Protocol Data Unit? Optional?

It probably shows my background, but I tend to name this kind of thing `maybe_$name` e.g. `if let Some(pdu) = maybe_pdu { ... }`, or I'll name it the name directly and shadow it in the binding, e.g. `if let Some(pdu) = pdu { ... }`. IOW: either something that's obviously descriptive of the optionality, or just shadow it.

File: statsrelay/src/backend_client/prom_client.rs [hunk 9]
```diff
@@ -0,0 +1,522 @@
...
+                } else {
+                    return (batch, true)
+                }
+            }
+            _ = &mut sleep => return (batch, false)
+        }
+    }
+}
+
+fn construct_write_request(stats: &Stats, batch: Vec<Pdu>) -> WriteRequest {
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:282]: This isn't something you need to do anything about, just something to keep in mind: this function allocates a bit.

My guess is it's completely fine, but if you find yourself needing to reduce allocator pressure or decrease peak heap, I'd look in here since there's a lot of `strcpy` behind the owned string construction.

File: statsrelay/src/backend_client/prom_client.rs [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+    #[allow(clippy::ptr_arg)]
+    fn decode_response(compressed: &Vec<u8>) -> WriteRequest {
+        let decompressed = snap::raw::Decoder::new()
+            .decompress_vec(compressed)
+            .expect("failed snappy decompression");
+        WriteRequest::decode(Cursor::new(decompressed)).expect("failed decoding write request")
+    }
+
+    #[tokio::test]
+    async fn test_prom_client() {
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:404]: This feels like a good candidate for quickcheck, if you have a structural representation of statsd pdus you can use for comparison. It's fine as is, with the by-hand test, but with protocol stuff like this I always feel a little more confident I've got all the edge cases when it's automated.

File: statsrelay-protobuf/proto/ext/github.com/prometheus/prometheus/prompb/types.proto [hunk 11]
```diff
@@ -44,7 +44,7 @@ message Sample {
   // conversion from time.Time to Prometheus timestamp.
   int64 timestamp = 2;
 
-  // added as an extension.
+  // added as an extension, sample_rate is a ratio in the range (0,1].
```
Comment (bearcage) [inline statsrelay-protobuf/proto/ext/github.com/prometheus/prometheus/prompb/types.proto:47]: Should the default be `1` rather than `100`?

File: statsrelay/src/backend_client/prom_client.rs [hunk 12]
```diff
@@ -213,23 +212,27 @@ async fn client_task<T: PromWriteRequestClient + Send + Sync + 'static>(
...
-                        new_res = client.send_write_request(compressed_write_request.clone()) => {
-                            res = new_res;
-                        },
-                        _ = trip.clone() => {
-                            break;
-                        },
-                    }
-                }
+                let res = retry_notify(
+                    ExponentialBackoff::default(),
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:216]: Nice!

File: statsrelay/src/backend_client/prom_client.rs (hunk 10 above)
Comment (dsmith3197) [reply statsrelay/src/backend_client/prom_client.rs]: I finally took the time to learn how to use quickcheck and it's great!

**Review by bearcage — APPROVED** [review]
This looks great to me -- thanks for taking the time with all those nits, you're elevating the code quality with this PR!

File: statsrelay/src/backend_client/prom_client.rs [hunk 13]
```diff
@@ -419,4 +434,30 @@ mod tests {
...
+            PromClient::new_with_client(scope, mock_client.clone(), endpoint, config.clone());
+
+        let pdu = Pdu::parse(Bytes::from_static(b"foo.bar:3|c|@1.0|#tags")).unwrap();
+
+        prom_client.try_send(pdu).expect("failed to send pdu");
+
+        let join_handle = prom_client.join_handle().unwrap();
+        drop(prom_client);
+        // This will block forever if we don't abort retries. 
+        let _ = join_handle.await;
```
Comment (bearcage) [inline statsrelay/src/backend_client/prom_client.rs:461]: What's the behavior of this failing in CI -- is it worth wrapping this in `tokio::time::timeout(...)` so to turn the infinite block into failure after X seconds so we don't e.g. wait for the github actions timeout to kill the run?

**Review by bearcage — APPROVED** [review]

File: statsrelay/src/backend_client/prom_client.rs (hunk 13 above)
Comment (dsmith3197) [reply statsrelay/src/backend_client/prom_client.rs]: Ah thanks for catching that. I added a timeout.

**Review by bearcage — APPROVED** [review]
